Improve next element computation in completion

msujew opened the pull request:
Closes https://github.com/eclipse-langium/langium/issues/1203

Improves the completion algorithm in two ways:
1. Correctly propagates the current `NextFeature#property` variable through alternatives (related to the linked issue)
2. Fixes the way in which we create new AST nodes during cross reference completion. Previously, the `new` property wasn't set correctly, which led to cross reference completion on "stale" elements. This change simplifies (and also improves) the computation by only using the `type` property.

Both changes aren't trivial to (automatically) test and require debugging to accurately pinpoint the actual changes in the completion algorithm.

msujew commented:
@cdietrich Are you interested in testing the change to see whether it works for you?

cdietrich commented:
@msujew i can have a look but i will need some time as i dont want to mess up my production dev env with npm link

cdietrich commented:
@msujew have problems to get this running

when i run my extension i am getting

```
Error: Cannot find module 'vscode'
Require stack:
- /Users/dietrich/tmp/viechviechviech/dot-expression-problem/out/language/main.cjs
    at Function._resolveFilename (node:internal/modules/cjs/loader:1129:15)
    at node:internal/modules/cjs/loader:974:27
    at Function._load (node:electron/js2c/asar_bundle:2:13377)
    at Module.require (node:internal/modules/cjs/loader:1195:19)
    at require (node:internal/modules/cjs/helpers:110:18)
    at node_modules/vscode-languageclient/lib/node/main.js (/Users/dietrich/tmp/viechviechviech/dot-expression-problem/out/language/main.cjs:35831:20)
    at __require (/Users/dietrich/tmp/viechviechviech/dot-expression-problem/out/language/main.cjs:26:50)
    at Object.<anonymous> (/Users/dietrich/tmp/viechviechviech/dot-expression-problem/out/language/main.cjs:63806:44)
    at Module._compile (node:internal/modules/cjs/loader:1316:14)
    at Object..js (node:internal/modules/cjs/loader:1371:10) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [
    '/Users/dietrich/tmp/viechviechviech/dot-expression-problem/out/language/main.cjs'
  ]
  ```
  any idea?

msujew commented:
@cdietrich Are you accidentally importing `vscode-languageclient` in your language server code? It imports `vscode` to register the language client handlers.

cdietrich commented:
yes.
i assume i have to import
import { CompletionItemKind, InsertTextFormat } from "vscode-languageserver"

cdietrich commented:
next.property is now correctly computed. thx.

sailingKieler commented on packages/langium/src/lsp/completion/completion-provider.ts:
```diff
@@ -381,28 +376,26 @@ export class DefaultCompletionProvider implements CompletionProvider {
-                });
+                for (const description of this.filterCrossReferences(context, scope.getAllElements())) {
+                    acceptor(context, this.createReferenceCompletionItem(description));
+                }
```
This is just a different way of formulating the same, correct? (except for the changed signature of `filterCrossReferences(...)`)

msujew replied:
Kind of yes. I just like the `for` loop more, since it's easier to debug.

sailingKieler commented on packages/langium/src/lsp/completion/completion-provider.ts:
```diff
@@ -381,28 +376,26 @@ export class DefaultCompletionProvider implements CompletionProvider {
         if (assignment && node) {
-            if (crossRef.type && (crossRef.new || node.$type !== crossRef.type)) {
+            if (crossRef.type) {
                 node = {
```
why is `node` re-initialized although it is already truthy?
Is this branch supposed to cover the case of assigning a (the first) property of an astNode not existing in the AST yet?
Under which circumstances is `crossRef.type` (un-)initialized?
The var name `crossRef` is a bit misleading as it actually denotes a `NextFeature` referring to a crossRef. Esp. `crossRef.type` is confusing, as it denotes `NextFeature.type`, not `CrossReference.type`.

Please add some comments here (and there).

msujew replied:
We sometimes have parser rules that look like this:

```
Ref: ref=[Element:ID];
```

Since we want the next element in our completion algorithm, we actually complete the `ref` assignment/cross reference although the current element cannot be of type `Ref`. In this case, the `NextFeature#type` property is set to `Ref` (since it's the new containing type), and we build a new synthetic element that can be used to compute the cross reference.

It's generally uninitialized when we don't enter a new parser rule during completion. I.e. like in the hello world grammar:

```
Hello |
```

Where the cross reference at `|` is part of the same parser rule as `Hello`.

sailingKieler replied:
Thanks a lot 👍

sailingKieler reviewed (changes requested):
First of all, although you stated that these changes are difficult to mechanically test, I still miss some unit tests, at least for documenting the expected behavior of the *follow-element-computation* algorithm. I would appreciate to have a few, e.g. for being able to easily debug the implementation in case of unexpected behaviors.

General question/remark:
* Is it intended that for terminal or datatype rules no completion is returned at all, according to `completionFor(...)`? I guess some hint on what can be expected as arguments in `completionFor(...)` would be very helpful.

Technical question:
* Why does `interpretStackToken` prefix the `nextFeatures` with the existing `stack`, and `findNextFeatures` drops that prefix, again? Do I overlook any evaluations of the prefixed `nextFeatures`?

Apart from my complaints you build quite impressive stuff! 🤩

msujew commented:
> First of all, although you stated that these changes are difficult to mechanically test, I still miss some unit tests, at least for documenting the expected behavior of the follow-element-computation algorithm. I would appreciate to have a few, e.g. for being able to easily debug the implementation in case of unexpected behaviors.

Not really in scope of this PR. The current follow element computation is so heavily intertwined with the workings of the completion provider that I would need to replicate half of what we do in the completion provider to build a test for the follow element computation.

Note that you can debug it using the existing completion provider tests.

> Is it intended that for terminal or datatype rules no completion is returned at all, according to completionFor(...)?

If the terminal or datatype rule is part of a cross reference, then we fall into the expected cross reference case. Otherwise, there's no reasonable way to assume the expected value of a terminal or datatype value, so we don't offer any completion for that.

> Why does interpretStackToken prefix the nextFeatures with the existing stack, and findNextFeatures drops that prefix, again? Do I overlook any evaluations of the prefixed nextFeatures?

We use the stack during the whole computation phase, since we might need to go up the stack once we encounter the end of a parser rule. Only the absolutely last step of the algorithm actually drops the stack again, since we only need the last elements of it for the completion provider.

I've reverted the breaking change and added a few more comments here and there 👍

msujew commented on packages/langium/src/lsp/completion/follow-element-computation.ts:
```diff
@@ -146,7 +139,11 @@ function findFirstFeaturesInternal(options: { next: NextFeature, cardinalities:
+            next: {
+                feature: e,
+                type,
+                property: next.property
```
Yes 👍

sailingKieler reviewed (approved):
Thanks for the improvements! 🙏
